openwisp_notifications/tests/test_selenium.py

```diff
@@ -84,75 +73,92 @@ def test_notification_dialog_open_button_visibility(self):
-            EC.visibility_of_element_located((By.ID, f'ow-{notification.id}'))
-        )
-        self.web_driver.find_element(By.ID, f'ow-{notification.id}').click()
-        WebDriverWait(self.web_driver, 10).until(
-            EC.visibility_of_element_located((By.CLASS_NAME, 'ow-dialog-notification'))
-        )
-        dialog = self.web_driver.find_element(By.CLASS_NAME, 'ow-dialog-notification')
+        self.find_element(By.ID, 'openwisp_notifications').click()
+        self.find_element(By.ID, f'ow-{notification.id}').click()
+        dialog = self.find_element(By.CLASS_NAME, 'ow-dialog-notification')
         # This confirms the button is hidden
         dialog.find_element(By.CSS_SELECTOR, '.ow-message-target-redirect.ow-hide')
 
+    def test_email_unsubscribe_page(self):
+        with self.subTest('Token is invalid'):
+            self.open(reverse('notifications:unsubscribe'))
+            self.assertEqual(
+                self.find_element(By.TAG_NAME, 'h2').text, 'Invalid or Expired Link'
+            )
+
+        with self.subTest('User unsubscribe with valid URL'):
+            unsubscribe_link = get_unsubscribe_url_for_user(self.admin, False)
+            self.open(unsubscribe_link)
+            self.wait_for_visibility(By.ID, 'subscribed-message')
+            self.wait_for_invisibility(By.ID, 'unsubscribed-message')
+            toggle_btn = self.find_element(By.ID, 'toggle-btn')
+            self.assertEqual(toggle_btn.text, 'Unsubscribe')
+            toggle_btn.click()
+            self.wait_for_visibility(By.ID, 'confirm-unsubscribed')
+            self.wait_for_invisibility(By.ID, 'confirm-subscribed')
+            self.assertEqual(self.find_element(By.ID, 'toggle-btn').text, 'Subscribe')
+
+        with self.subTest('User subscribe to notifications again'):
+            self.open(unsubscribe_link)
+            self.wait_for_visibility(By.ID, 'unsubscribed-message')
+            self.wait_for_invisibility(By.ID, 'subscribed-message')
+            toggle_btn = self.find_element(By.ID, 'toggle-btn')
+            self.assertEqual(toggle_btn.text, 'Subscribe')
+            toggle_btn.click()
+            self.wait_for_visibility(By.ID, 'confirm-subscribed')
+            self.wait_for_invisibility(By.ID, 'confirm-unsubscribed')
+            self.assertEqual(self.find_element(By.ID, 'toggle-btn').text, 'Unsubscribe')
+
+        with self.subTest('Network request fails'):
+            self.open(unsubscribe_link)
+            self.web_driver.execute_script(
+                """
+                window.fetch = function() {
+                    return Promise.reject(new Error('Simulated fetch failure'));
+                };
+            """
+            )
+            self.web_driver.find_element(By.ID, 'toggle-btn').click()
+            self.wait_for_visibility(By.ID, 'error-msg')
+
     def test_notification_preference_page(self):
         self.login()
-        self.open('/notifications/preferences/')
```

Comment:
I've seen other occurrences of this, please look for `/notifications/` and update all occurrences.